File: packages/shared/src/schemas/course.ts
import { z } from "zod";

import { paginationQuerySchema } from "./pagination";

export const courseStatusSchema = z.enum(["DRAFT", "PUBLISHED", "ARCHIVED"]);

export const pricingTypeSchema = z.enum(["FREE", "PAID"]);

export const courseLevelSchema = z.enum([
  "BEGINNER",
  "INTERMEDIATE",
  "ADVANCED",
  "ALL_LEVELS",
]);

export const publicCoursesQuerySchema = paginationQuerySchema.extend({
  categorySlug: z.string().trim().min(1).optional(),
  search: z.string().trim().min(1).optional(),
});

export const adminCoursesQuerySchema = paginationQuerySchema.extend({
  search: z.string().trim().optional(),
  status: courseStatusSchema.optional(),
  pricingType: pricingTypeSchema.optional(),
});

export const courseSlugParamsSchema = z.object({
  slug: z
    .string()
    .trim()
    .min(1)
    .regex(/^[\p{L}\p{N}_\-]+$/u, "مسار الكورس غير صالح."),
});

export const courseIdParamsSchema = z.object({
  id: z.string().cuid(),
});

export const courseCreateBodySchema = z
  .object({
    title: z.string().trim().min(3, "العنوان قصير جدًا."),
    slug: z
      .string()
      .trim()
      .regex(/^[\p{L}\p{N}_\-]+$/u, "المعرّف غير صالح.")
      .optional(),
    description: z.string().trim().min(10, "الوصف مطلوب."),
    shortDescription: z.string().trim().max(500).optional(),
    thumbnailUrl: z.union([z.string().url(), z.literal("")]).optional(),
    categoryId: z.preprocess(
      (v) => (v === "" ? null : v),
      z.union([z.string().cuid(), z.null()]).optional(),
    ),
    pricingType: pricingTypeSchema,
    priceAmount: z.coerce.number().positive().optional(),
    currency: z.string().trim().length(3).default("JOD"),
    level: courseLevelSchema.default("ALL_LEVELS"),
    estimatedDurationMinutes: z.coerce.number().int().min(0).max(100000).optional(),
    status: courseStatusSchema.default("DRAFT"),
  })
  .superRefine((data, ctx) => {
    if (data.pricingType === "PAID") {
      if (data.priceAmount === undefined || Number.isNaN(data.priceAmount)) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "المبلغ مطلوب للكورسات المدفوعة.",
          path: ["priceAmount"],
        });
      }
    }
  });

export type CourseCreateBody = z.infer<typeof courseCreateBodySchema>;

export const courseUpdateBodySchema = z.object({
  title: z.string().trim().min(3).optional(),
  slug: z
    .string()
    .trim()
    .regex(/^[\p{L}\p{N}_\-]+$/u)
    .optional(),
  description: z.string().trim().min(10).optional(),
  shortDescription: z.string().trim().max(500).nullable().optional(),
  thumbnailUrl: z
    .union([z.string().url(), z.literal("")])
    .nullable()
    .optional(),
  categoryId: z.preprocess(
    (v) => (v === "" ? null : v),
    z.union([z.string().cuid(), z.null()]).optional(),
  ),
  pricingType: pricingTypeSchema.optional(),
  priceAmount: z.coerce.number().positive().nullable().optional(),
  currency: z.string().trim().length(3).optional(),
  level: courseLevelSchema.optional(),
  estimatedDurationMinutes: z
    .coerce.number()
    .int()
    .min(0)
    .max(100000)
    .nullable()
    .optional(),
  status: courseStatusSchema.optional(),
});

export type CourseUpdateBody = z.infer<typeof courseUpdateBodySchema>;

File: apps/mobile/lib/src/features/courses/models/course.dart
import "category.dart";

class Course {
  const Course({
    required this.id,
    required this.title,
    required this.slug,
    this.shortDescription,
    this.description,
    this.thumbnailUrl,
    required this.pricingType,
    this.priceAmount,
    required this.currency,
    required this.level,
    this.estimatedDurationMinutes,
    this.publishedAt,
    this.category,
    this.lessonCount = 0,
    this.appleProductId,
    this.iosPurchasable = false,
    this.apiIncludesIapFields = false,
  });

  final String id;
  final String title;
  final String slug;
  final String? shortDescription;
  final String? description;
  final String? thumbnailUrl;
  final String pricingType;
  final String? priceAmount;
  final String currency;
  final String level;
  final int? estimatedDurationMinutes;
  final String? publishedAt;
  final Category? category;
  final int lessonCount;
  final String? appleProductId;
  final bool iosPurchasable;
  final bool apiIncludesIapFields;

  bool get isFree => pricingType == "FREE";

  bool get isIosIapPurchasable =>
      !isFree &&
      iosPurchasable &&
      (appleProductId?.trim().isNotEmpty ?? false);

  String get priceLabel => isFree ? "مجاني" : "${priceAmount ?? "—"} $currency";

  String get levelLabel => switch (level) {
    "BEGINNER" => "مبتدئ",
    "INTERMEDIATE" => "متوسط",
    "ADVANCED" => "متقدم",
    _ => "جميع المستويات",
  };

  factory Course.fromJson(Map<String, dynamic> json) {
    final cat = json["category"];
    return Course(
      id: json["id"] as String,
      title: json["title"] as String,
      slug: json["slug"] as String,
      shortDescription: json["shortDescription"] as String?,
      description: json["description"] as String?,
      thumbnailUrl: json["thumbnailUrl"] as String?,
      pricingType: json["pricingType"] as String? ?? "FREE",
      priceAmount: json["priceAmount"]?.toString(),
      currency: json["currency"] as String? ?? "JOD",
      level: json["level"] as String? ?? "ALL_LEVELS",
      estimatedDurationMinutes: (json["estimatedDurationMinutes"] as num?)
          ?.toInt(),
      publishedAt: json["publishedAt"] as String?,
      category: cat is Map<String, dynamic> ? Category.fromJson(cat) : null,
      lessonCount: (json["lessonCount"] as num?)?.toInt() ?? 0,
      appleProductId: json["appleProductId"] as String?,
      iosPurchasable: json["iosPurchasable"] as bool? ?? false,
      apiIncludesIapFields:
          json.containsKey("iosPurchasable") ||
          json.containsKey("appleProductId"),
    );
  }
}
